cider/function/substrait/SubstraitExtension.cpp

```diff
@@ -0,0 +1,349 @@
+        for (auto& impl : impls) {
+          auto aggregateFunctionVariant = impl.as<SubstraitAggregateFunctionVariant>();
+          aggregateFunctionVariant.name = function.name;
+          function.impls.emplace_back(std::make_shared<SubstraitAggregateFunctionVariant>(
+              aggregateFunctionVariant));
+        }
+      }
+      return true;
+    }
+    return false;
+  }
+};
+
+template <>
+struct convert<cider::function::substrait::SubstraitTypeAnchor> {
+  static bool decode(const Node& node,
+                     cider::function::substrait::SubstraitTypeAnchor& typeAnchor) {
+    auto& name = node["name"];
+    if (name && name.IsScalar()) {
+      typeAnchor.name = name.as<std::string>();
+      return true;
+    }
+    return false;
+  }
+};
+
+template <>
+struct convert<cider::function::substrait::SubstraitExtension> {
+  static bool decode(const Node& node,
+                     cider::function::substrait::SubstraitExtension& extension) {
+    auto& scalarFunctions = node["scalar_functions"];
+    auto& aggregateFunctions = node["aggregate_functions"];
+    const bool scalarFunctionsExists = scalarFunctions && scalarFunctions.IsSequence();
+    const bool aggregateFunctionsExists =
+        aggregateFunctions && aggregateFunctions.IsSequence();
+    if (!scalarFunctionsExists && !aggregateFunctionsExists) {
+      return false;
+    }
+
+    if (scalarFunctionsExists) {
+      for (auto& scalarFunctionNode : scalarFunctions) {
+        const auto& scalarFunction = scalarFunctionNode.as<SubstraitScalarFunction>();
+        for (auto& scalaFunctionVariant : scalarFunction.impls) {
+          extension.scalarFunctionVariants.emplace_back(scalaFunctionVariant);
+        }
+      }
+    }
+
+    if (aggregateFunctionsExists) {
+      for (auto& aggregateFunctionNode : aggregateFunctions) {
+        const auto& aggregateFunction =
+            aggregateFunctionNode.as<SubstraitAggregateFunction>();
+        for (auto& aggregateFunctionVariant : aggregateFunction.impls) {
+          extension.aggregateFunctionVariants.emplace_back(aggregateFunctionVariant);
+        }
+      }
+    }
+
+    auto& types = node["types"];
+    if (types && types.IsSequence()) {
+      for (auto& type : types) {
+        auto typeAnchor = type.as<SubstraitTypeAnchor>();
+        extension.types.emplace_back(std::make_shared<SubstraitTypeAnchor>(typeAnchor));
+      }
+    }
+
+    return true;
+  }
+};
+
+}  // namespace YAML
+
+namespace cider::function::substrait {
+
+namespace {
+
+std::string getSubstraitExtensionAbsolutePath() {
+  const std::string absolute_path = __FILE__;
+  auto const pos = absolute_path.find_last_of('/');
+  return absolute_path.substr(0, pos) + "/extensions/";
+}
+
+}  // namespace
+
+std::shared_ptr<SubstraitExtension> SubstraitExtension::loadExtension() {
+  static const auto& extension = loadDefault();
+  return extension;
+}
+
+std::shared_ptr<SubstraitExtension> SubstraitExtension::loadDefault() {
```

Comment:
Do we need to call methods like this one? if not, you can remove it to avoid code redundancy.

Reply:
loadDefault contains the yaml file path such as: "functions_arithmetic.yaml", "functions_arithmetic_decimal.yaml", "functions_comparison.yaml", "unknown.yaml". So we should keep it.